src/main/java/com/rabbitmq/perf/Producer.java

```diff
@@ -154,33 +167,46 @@ public void run() {
         }
     }
 
-    private void publish(byte[] msg)
+    private void publish(BasicProperties basicProperties, byte[] msg)
         throws IOException {
 
         unconfirmedSet.add(channel.getNextPublishSeqNo());
         channel.basicPublish(exchangeName, randomRoutingKey ? UUID.randomUUID().toString() : id,
                              mandatory, immediate,
-                             persistent ? MessageProperties.MINIMAL_PERSISTENT_BASIC : MessageProperties.MINIMAL_BASIC,
-                             msg);
+                             basicProperties, msg);
     }
 
-    private byte[] createMessage(int sequenceNumber)
+    private BasicProperties createProperties(int sequenceNumber, long nano) {
+        if (!setProperties) {
+            return persistent ? MessageProperties.MINIMAL_PERSISTENT_BASIC : MessageProperties.MINIMAL_BASIC;
+        }
+        return new BasicProperties.Builder()
+                .appId(this.getClass().getName())
+                .contentType("application/octet-stream")
+                .deliveryMode(persistent ? 2 : 1)
+                .messageId(Integer.toString(sequenceNumber))
+                .priority(0)
+                .timestamp(new Date(nano/1000000))
```

Comment:
Using `System.currentTimeMillis` would make more sense if you want a date and a time. `System.nanoTime` is meant to measure elapsed time, you can end up with funny results by using it with `Date` (even negative values, as stated in the javadoc).
Regarding `priority(0)`, same question as @michaelklishin.